postgresqlPackages.pgvecto-rs: init at 0.2.1

diogotcorreia opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Description: Scalable Vector Search in Postgres. Revolutionize Vector Search, not Database. 
https://github.com/tensorchord/pgvecto.rs

Used for [Immich](https://github.com/immich-app/immich).

Closes #274509

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

~~I have not tested this with Immich before, but as per https://github.com/NixOS/nixpkgs/issues/274509#issuecomment-1892563516, it seems like this version would not work (because of upstream).~~

~~Keeping this as a draft until Immich moves to a more recent pgvecto.rs version.~~

Edit: see https://github.com/NixOS/nixpkgs/pull/281192#issuecomment-1937355090 (next version of Immich will use pgvecto.rs v0.2.0)

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

diogotcorreia committed:
pgvecto.rs: init at 0.2.1

Co-Authored-By: Daniel Albert <git@esclear.de>
Co-Authored-By: rina <k@rina.fyi>

diogotcorreia commented:
It looks like v0.2.0 has been released and that Immich is moving to that version of pgvecto.rs in the next release: https://github.com/immich-app/immich/pull/6785

I'll be working tomorrow to get this ready in time for the release!

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,77 @@
+}:
+
+let
+  cargo-pgrx_0_11_2 = cargo-pgrx.overrideAttrs (old: rec {
```
Either way would be acceptable; we don't know how well it'd take newer versions. I'd try to make the original Nixpkgs package work and make it capable of multi-version if push comes to shove but this is fine too.

Either way, add this info as a code comment.

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,77 @@
+, cargo-pgrx
+, fetchCrate
+, openssl
+, clang
```
I don't know this section of Nixpkgs very well but, reading the source a bit, it appears you can create your own `rustPlatform` using `makeRustPlatform` and pass it your desired version of `clang`.

i.e.

```Nix
makeRustPlatform {
  clang = clang_16;
  inherit rustc cargo;
}
```

https://nixos.org/manual/nixpkgs/stable/#using-rust-nightly-in-a-derivation-with-buildrustpackage

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,78 @@
+    hash = "sha256-30AzS9R01+Ntq/HKabn+3tRHiEe68pxoVK3sbr+wFWg=";
+  };
+
+  # package has git dependencies on Cargo.lock, so cargoHash does not work
```
Not familiar with rust but what exactly does this mean?

diogotcorreia replied:
This means they have at least one dependency that is fetched from a git repo instead of from crates.io.
https://github.com/tensorchord/pgvecto.rs/blob/be5a816810f63cadada90a666cfc70db37c78a33/Cargo.toml#L35

I can try to clarify more in the comment and put there the link above as well. I took a look at the master branch now and that dependency is gone, so it seems this might stop being a problem in the future.

Atemu reviewed (commented):
This already looks quite good. A couple minor things.

I'd also like to know how you test this works as expected.

Atemu commented on pkgs/servers/sql/postgresql/packages.nix:
```diff
@@ -44,6 +44,8 @@ self: super: {
 
     pgsql-http = super.callPackage ./ext/pgsql-http.nix { };
 
+    pgvecto-rs = super.callPackage ./ext/pgvecto-rs { };
+
```
@infinisil by-name CI fails here but I don't think it should since this isn't all-packages.nix.

diogotcorreia replied:
It only started failing after I edited the PR description last night btw
I'll also take a look to see if it's something I can fix on my side, but it seems to be a problem with the CI at first glance

diogotcorreia replied:
Seems to have been caused by commit https://github.com/NixOS/nixpkgs/commit/a8a7369c36120acb9ca43376504b6e9babed7261, which was reverted in the meantime, https://github.com/NixOS/nixpkgs/commit/ac637ef21ecc2f0a1a0987fe9ce1d7f0d7f42fca.
I've edited the PR description again to trigger the CI and it passes now.

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,96 @@
+  # avoid future incompatibility.
+  # See https://docs.pgvecto.rs/developers/development.html#environment, step 4
+  rustPlatform' = rustPlatform // {
+    inherit (callPackage ../../../../../build-support/rust/hooks {
```
Relative paths are not part of a package's interface. We'll have to find a different solution.

diogotcorreia replied:
You're absolutely right. I'll investigate alternative solutions

diogotcorreia replied:
There doesn't seem to be a way to override it. This is what I've tried:

```
nix-repl> pkgs.rustPlatform.bindgenHook.override { clang = pkgs.clang_17; }
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at «string»:1:1:

            1| pkgs.rustPlatform.bindgenHook.override { clang = pkgs.clang_17; }
             | ^

       … while calling a functor (an attribute set with a '__functor' attribute)

         at /home/dtc/documents/vcs/nixpkgs/lib/customisation.nix:104:43:

          103|       # Re-call the function but with different arguments
          104|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                           ^
          105|       # Change the result of the function call by applying g to it

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'clang'

       at /home/dtc/documents/vcs/nixpkgs/pkgs/build-support/rust/hooks/default.nix:92:32:

           91|
           92|     bindgenHook = callPackage ({}: makeSetupHook {
             |                                ^
           93|       name = "rust-bindgen-hook";

```

```
nix-repl> (pkgs.makeRustPlatform {rustc = pkgs.rustc; cargo = pkgs.cargo; clang = pkgs.clang_17;}).bindgenHook.clang
«derivation /nix/store/j16048gsic45ijdpw82172pfxdz9nw2w-clang-wrapper-16.0.6.drv»
```

For now, I'll setup a similar script just for this package that supports pinning the clang version, but this should be revisited at a later date to allow overriding this.

Atemu replied:
This looks like a bug. It can be fixed with this:

```patch
diff --git a/pkgs/development/compilers/rust/make-rust-platform.nix b/pkgs/development/compilers/rust/make-rust-platform.nix
index e22cb6d594af..6ed724aae821 100644
--- a/pkgs/development/compilers/rust/make-rust-platform.nix
+++ b/pkgs/development/compilers/rust/make-rust-platform.nix
@@ -1,4 +1,4 @@
-{ lib, buildPackages, callPackage, cargo-auditable, stdenv, runCommand }@prev:
+{ lib, buildPackages, callPackage, callPackages, cargo-auditable, stdenv, runCommand }@prev:
 
 { rustc
 , cargo
@@ -34,7 +34,7 @@ rec {
   };
 
   # Hooks
-  inherit (callPackage ../../../build-support/rust/hooks {
+  inherit (callPackages ../../../build-support/rust/hooks {
     inherit stdenv cargo rustc;
   }) cargoBuildHook cargoCheckHook cargoInstallHook cargoNextestHook cargoSetupHook maturinBuildHook bindgenHook;
 }
```

diogotcorreia replied:
I'll investigate that later then, thanks :)
Should this be made in another PR, or does it make sense to include it in this one?

diogotcorreia replied:
Great, it seems to work fine in the repl. I've tested it by overriding clang to 17 and having it fail the build.
I'm just finishing building the package to make sure everything works fine, and then I'll push.
This will likely trigger a lot of package rebuilds in the CI, but we'll see :upside_down_face:

PhilippWoelfel replied:
> Relative paths are not part of a package's interface. We'll have to find a different solution.

Sorry, I didn't understand that comment, and I'm curious about this... Why is the relative path a problem? Also, for example, `pkgs/os-specific/darwin/apple-sdk-11.0/default.nix` does something similar:
```
    rustPlatform = pkgs.makeRustPlatform {
      inherit (pkgs.darwin.apple_sdk_11_0) stdenv;
      inherit (pkgs) rustc cargo;
    } // {
      inherit (pkgs.callPackage ../../../build-support/rust/hooks {
        inherit (pkgs.darwin.apple_sdk_11_0) stdenv;
        inherit (pkgs) cargo rustc;
        clang = mkCc pkgs.clang;
      }) bindgenHook;
    };
```

Atemu replied:
Relative paths are not guaranteed to remain stable.

People who reorganise one file in nixpkgs should not be expected to check and adjust relative paths in all other files.

This is relevant as we'll soon see the great RFC140 migration.

Now that the hooks override is fixed, the the apple SDK should use this style aswell as it doesn't have this obvious issue. (It may still have less obvious issues.)

diogotcorreia committed:
nixos/tests/pgvecto-rs: init

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,107 @@
+
+  # Set appropriate version on vectors.control, otherwise it won't show up on PostgreSQL
+  postPatch = ''
+    sed -i "s/@CARGO_VERSION@/${version}/g" ./vectors.control
```
```suggestion
    substituteInPlace ./vectors.control --subst-var-by CARGO_VERSION ${version}
```

PhilippWoelfel commented:
Can you add the source files from `sql/upgrade` to the output? E.g.,
```
  postInstall = ''
    cp sql/upgrade/* $out/share/postgresql/extension/
  '';
```

diogotcorreia commented:
@PhilippWoelfel that was also in my todo list as well, yes :)
I was still going to investigate how they do it on their CI, since the normal build process doesn't seem to include them.

I'm not sure if I'm going to be able to finish the remaining problems this with PR this weekend, but I'll do my best :sweat_smile:

diogotcorreia commented:
I think it should be ready now :) :crossed_fingers: 

Changes I've made:
- Check extension version in the tests
- Properly pinned clang 16 by setting an env variable in `preBuild`
- Changed env variable name in patch 0001 to `C_CRATE_EXTRA_CLANG_ARGS` to avoid conflict with `rustPlatform.bindgenHook`
- Replaced use of `sed` with `substituteInPlace` as per the review
- Included the upgrade SQL scripts (thanks @PhilippWoelfel !)
- Updated package description as per upstream

Since I'm force pushing, here's the diff: https://bin.diogotc.com/xuzecylyju.diff

Atemu commented:
Re: force pushing obfuscating github logs, you could just push them as simple commits and squash/fixup later. If you had to rebase, force-push the rebase of the original commits first and then the additional commits and/or squash/fixup.

That way changes remain traceable.

diogotcorreia committed:
rustPlatform: fix `override` on hooks

Previously, trying to use `.override {}` on one of the hooks,
specifically `bindgenHook` would yield in an error.
By replacing `callPackage` with `callPackages`, this error is fixed and
the overrides are propagated to the hooks.

Co-Authored-By: Atemu <atemu.main+nixpkgs@gmail.com>

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -9,13 +9,17 @@
   # avoid future incompatibility.
   # See https://docs.pgvecto.rs/developers/development.html#environment, step 4
   clang = clang_16;
+  rustPlatform' = rustPlatform // {
+    bindgenHook = rustPlatform.bindgenHook.override { inherit clang; };
+  };
```
I'd prefer a full `makeRustPlatform` call.

diogotcorreia replied:
I tried that, but it seems to not be overriding as expected.
```
nix-repl> (pkgs.makeRustPlatform {rustc = pkgs.rustc; cargo = pkgs.cargo; clang = pkgs.clang_17;}).bindgenHook.clang
«derivation /nix/store/j16048gsic45ijdpw82172pfxdz9nw2w-clang-wrapper-16.0.6.drv»
```

Am I doing it wrong?

Atemu replied:
Ah, `makeRustPlatform` doesn't take a `clang` arg but does accept arbitrary args, so it doesn't throw an error to do this.

I don't see another option short of refactoring `makeRustPlatform` then.

I really don't know how this all works together; doesn't rust internally depend on clang aswell? Shouldnt'y you need to override that aswell?

I'd like input of the rust people on this. It appears they were pinged already.

Mic92 replied:
This is what we do for firefox: https://github.com/NixOS/nixpkgs/blob/9d422aeafc8780e1efe49f5891b70685d9c05c40/pkgs/applications/networking/browsers/firefox/common.nix#L177

Mic92 replied:
Not sure if it also applies to this.

diogotcorreia replied:
@Mic92 Thanks for you help, but it seems that it still doesn't work. Let me know if I'm using this wrong:
```
nix-repl> (pkgs.makeRustPlatform { inherit (pkgs) rustc cargo; stdenv = pkgs.overrideCC pkgs.llvmPackages.stdenv (llvmPackages.stdenv.override { cc = pkgs.clang_17; }); }).bindgenHook.clang
«derivation /nix/store/c9ym6jk1pf0bx62j29g51k00pp16zf9l-clang-wrapper-16.0.6.drv»

nix-repl> (pkgs.makeRustPlatform { inherit (pkgs) rustc cargo; stdenv = pkgs.overrideCC pkgs.llvmPackages.stdenv.override { cc = pkgs.clang_17; }; }).bindgenHook.clang
«derivation /nix/store/c9ym6jk1pf0bx62j29g51k00pp16zf9l-clang-wrapper-16.0.6.drv»

nix-repl> (pkgs.makeRustPlatform { inherit (pkgs) rustc cargo; stdenv = pkgs.llvmPackages.stdenv.override { cc = pkgs.clang_17; }; }).bindgenHook.clang
«derivation /nix/store/c9ym6jk1pf0bx62j29g51k00pp16zf9l-clang-wrapper-16.0.6.drv»
```

For context, this is what I want to achieve (this is the current solution in the code):
```
nix-repl> (pkgs.rustPlatform.bindgenHook.override { clang = pkgs.clang_17; }).clang
«derivation /nix/store/cxxk9njrx1zm5ynmsi0wyy9dv58h2zci-clang-wrapper-17.0.6.drv»
```

(Also, I do want clang 16, but I want to pin it; I'm just using clang 17 in the repl to make sure it is being overridden)

Is there something I'm overlooking?

diogotcorreia commented:
Result of `nixpkgs-review pr 281192` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>postgresql12JitPackages.pgvecto-rs</li>
    <li>postgresql12Packages.pgvecto-rs</li>
    <li>postgresql13JitPackages.pgvecto-rs</li>
    <li>postgresql13Packages.pgvecto-rs</li>
    <li>postgresql14JitPackages.pgvecto-rs</li>
    <li>postgresql14Packages.pgvecto-rs</li>
    <li>postgresqlJitPackages.pgvecto-rs (postgresql15JitPackages.pgvecto-rs)</li>
    <li>postgresqlPackages.pgvecto-rs (postgresql15Packages.pgvecto-rs)</li>
    <li>postgresql16JitPackages.pgvecto-rs</li>
    <li>postgresql16Packages.pgvecto-rs</li>
  </ul>
</details>

Atemu commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/0001-read-clang-flags-from-environment.diff:
```diff
@@ -0,0 +1,17 @@
+     cc::Build::new()
+         .compiler("clang-16")
+         .file("./src/c.c")
++        .try_flags_from_environment("BINDGEN_EXTRA_CLANG_ARGS")
```
Could you add that fact as an explicit comment?

diogotcorreia replied:
Done, and squashed/reordered the commits so that it's ready to merged.

Atemu reviewed (approved):
I don't see any obvious issues on this anymore. I'll wait a few days for input from rust people on how to properly override clang in the rust platform but merge after that.

diogotcorreia commented:
Final testing remark: I am now using this in my Immich instance and it seems to be running fine :partying_face: 

https://github.com/diogotcorreia/dotfiles/commit/70e71c4a8160a3cf0132a8553b086b29bae749dd

Atemu reviewed (approved):
When a better solution to the rustPlatform clang issue is found, we can always do that later.

Apart from that, this PR has no obvious issues, so I'm inclined to merge.

Atemu commented:
The aarch64-linux and all Darwin builds appears to be broken. Could you mark them as such?

diogotcorreia commented:
@Atemu Should I use `meta.badPlatforms` or `meta.broken`?

Atemu commented:
Just `meta.broken`. I don't think I've ever see `badPlatforms` used for anything and if a build will not produce the expected result, `meta.broken` it is.

Atemu commented:
https://github.com/ofborg build postgresqlPackages.pgvecto-rs postgresqlPackages.pgvecto-rs.passthru.tests

marsam commented:
Hi, thanks a lot for your contribution. Sorry for the delay.
Closing in favor of https://github.com/NixOS/nixpkgs/pull/287632 I've integrated your changes there. Thanks!

diogotcorreia commented:
@marsam 
Hey!
I don't want to come across as rude, but I was surprised to see this PR get closed given that is was ready to merge and been around for longer than the other one (and also the one mentioned in the packaging request issue).
I'm fairly new to contributing to Nixpkgs, so I'm not sure how things work here -- **is there any way to avoid these kinds of collisions in the future?** Before starting a PR, I usually make sure that there is no existing issue/PR of the same thing.

Ultimately, what matters is that this gets added to Nixpkgs and I can't wait for either of the PRs to be merged! :) But it is fairly discouraging to see many of the hours I put into this PR being "thrown out", as you sure understand.

Sorry for the small "rant", and let me know if something is needed on the other PR (it, for example, doesn't seem to pin `cargo-pgrx` like mine does; not sure if something else is missing).

marsam commented:
Not rude at all; I would word it the same way you did.

I'm so sorry, I should have manage this better, it's totally my fault. It was not nice from me to make this arbitrary decision without consulting you first.

Why did I choose the other PR? somehow, I just had it in my notifications first, and did a quick test. I noticed your PR after; so I thought we could merge your changes in the other PR. It was an arbitrary decision, and I handled it poorly.
In retrospective I should have promoted you PR instead.

I apologize for my behavior and I hope this doesn't discourage you to keep on contributing to nixpkgs.
I'll reflect on how to handle this better in the future.

diogotcorreia commented:
@marsam Thank you for your reply, I understand that handling so many PRs can be overwhelming :sweat_smile: Thanks for all the work :)
I will for sure keep contributing to nixpkgs :rocket: 

Onto more practical matters: if you think this PR is in a better state than the other one (i.e., it is basically ready to merge), I don't mind updating it to 0.2.1 so that we can get it merged :)
And obviously, I can add @esclear as a maintainer as well, if they so desire.

marsam commented:
Sure, I think we can reopen this one :)

marsam commented on nixos/tests/pgvecto-rs.nix:
```diff
@@ -0,0 +1,76 @@
+  make-postgresql-test = postgresql-name: postgresql-package: makeTest {
+    name = postgresql-name;
+    meta = with pkgs.lib.maintainers; {
+      maintainers = [ typetetris ];
```
```suggestion
      maintainers = [ diogotcorreia ];
```

Atemu commented on nixos/tests/pgvecto-rs.nix:
```diff
@@ -0,0 +1,76 @@
+    '';
+
+  };
+  applicablePostgresqlVersions = filterAttrs (_: value: versionAtLeast value.version "12") postgresql-versions;
```
This may change in the future, so I think it's fine to keep it.

diogotcorreia commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,113 @@
+  # See https://docs.pgvecto.rs/developers/development.html#environment, step 4
+  clang = clang_16;
+  rustPlatform' = rustPlatform // {
+    bindgenHook = rustPlatform.bindgenHook.override { inherit clang; };
```
Just tried this out with `clang_17` and it still built, so it means it doesn't work :(

Atemu commented:
@ofBorg build postgresqlPackages.pgvecto-rs, postgresqlPackages.pgvecto-rs.passthru.tests

The CI failure is unrelated, you can ignore it.

diogotcorreia commented:
Just deployed this on my Immich server, everything working perfectly!

Atemu reviewed (approved):
I think this is GTG for an initial implementation.

Could you squash the fixup commits?

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,119 @@
+
+  meta = with lib; {
+    broken = (stdenv.isLinux && stdenv.isAarch64) || stdenv.isDarwin;
+    description =
+      "Scalable, Low-latency and Hybrid-enabled Vector Search in Postgres. Revolutionize Vector Search, not Database.";
```
```suggestion
    description = "Scalable, Low-latency and Hybrid-enabled Vector Search in Postgres";
```
This should just be a half sentence and not two sentences

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,119 @@
+  };
+
+  meta = with lib; {
+    broken = (stdenv.isLinux && stdenv.isAarch64) || stdenv.isDarwin;
```
Can we add a short comment why this isn't working or set platforms to just x86_64-linux?

diogotcorreia replied:
I think @Atemu advocated for `broken` instead of `platforms` here: https://github.com/NixOS/nixpkgs/pull/287632#discussion_r1513851890

As for why it's broken in linux_aarch64, I personally have no idea as I personally don't have any ARM machine to try it out. Judging from [comments in the other PR](https://github.com/NixOS/nixpkgs/pull/287632#issuecomment-1971770586), it seems to be an issue with bindgen that has already been resolved, but pgvecto-rs hasn't incorporated the changes yet; I'll that as a note.
As for darwin, I don't see any official builds from pgvecto-rs, but darwin is listed in the `targets` section of the [`rust-toolchain.toml`](https://github.com/tensorchord/pgvecto.rs/blob/875a20e412f041c2145ce698d6cc545ab0847ec2/rust-toolchain.toml) file upstream, so _in theory_ it should work. Again, I don't own any `darwin` device, so I can't really test this.

diogotcorreia replied:
I'll update the comment with the issue link for pgrx since it's more relevant, thanks!

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,119 @@
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "openai_api_rust-0.1.8" =
+        "sha256-os5Y8KIWXJEYEcNzzT57wFPpEXdZ2Uy9W3j5+hJhhR4=";
+      "std_detect-0.1.5" =
+        "sha256-RwWejfqyGOaeU9zWM4fbb/hiO1wMpxYPKEjLO0rtRmU=";
```
```suggestion
      "openai_api_rust-0.1.8" = "sha256-os5Y8KIWXJEYEcNzzT57wFPpEXdZ2Uy9W3j5+hJhhR4=";
      "std_detect-0.1.5" = "sha256-RwWejfqyGOaeU9zWM4fbb/hiO1wMpxYPKEjLO0rtRmU=";
```

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,119 @@
+  # so we're pinning it here to avoid future incompatibility.
+  # See https://docs.pgvecto.rs/developers/development.html#environment, step 6
+  cargo-pgrx_0_11_2 = cargo-pgrx.overrideAttrs (old: rec {
+    name = "cargo-pgrx-${version}";
```
```suggestion
    pname = "cargo-pgrx";
```

diogotcorreia commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,112 @@
+  inherit postgresql;
+
+  pname = "pgvecto-rs";
+  version = "0.2.1";
```
Done, thanks for noticing!

diogotcorreia commented on pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:
```diff
@@ -0,0 +1,118 @@
+    description = "Scalable, Low-latency and Hybrid-enabled Vector Search in Postgres";
+    homepage = "https://github.com/tensorchord/pgvecto.rs";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ diogotcorreia ];
```
@esclear Do you want to be added as a maintainer here as well?

marsam reviewed (approved):
LGTM. Thanks a lot!

diogotcorreia commented:
Just added @esclear as maintainer as well, as requested.

Hopefully that's all :rocket:

Atemu commented:
Thanks!
